### PR TITLE
Adds Shell Flyout Footer Template

### LINKED ISSUE #6292: [Spec]  Shell Flyout Footer Template
# Flyout Footer Template

Add a template for the footer of the flyout menu just like you have for the header. 

* The footer should be fixed to the bottom of the flyout. 
* The menu list should not go behind it and obscure the menu items.
* The footer should take any height.

```
<Shell.FlyoutFooterTemplate>
    <DataTemplate>
        <Grid>ContentHere</Grid>
    </DataTemplate>
</Shell.FlyoutFooterTemplate>
```

# API

Shell.FlyoutFooterTemplate takes a DataTemplate
 
# Difficulty : Low

### PR DESCRIPTION
**Pull request by ahoefling**
### Description of Change ###
Adds the Shell Flyout Footer Template implementation for Android meeting the following specs documented in the issue:

**[Spec]**
> - The footer should be fixed to the bottom of the flyout.
> - The menu list should not go behind it and obscure the menu items.
> - The footer should take any height.

 
### Issues Resolved ### 

- fix #6292

#### IShellController

Added:
```c#
View FlyoutFooter { get; }
```

#### Shell

Added:
```c#
View IShellController.FlyoutFooter { get; }
object FlyoutFooter { get; set; } // Bindable Property
DataTemplate FlyoutFooterTemplate { get; set; } // Bindable Property
```

There are several non-public methods added to this class, but the additions above are all the public changes that mirror `FooterHeader`

### Platforms Affected ### 

- Core/XAML (all platforms)
- Android

### Behavioral/Visual Changes ###
Changed the `ShellFlyoutTemplatedContentRenderer` to render an additional `ViewGroup` that is the fixed `FlyoutFooter` which appears at the bottom of the Flyout.

#### FlyoutContent.axml

Added: 
* `FrameLayout` which is used to add the content of `FlyoutFooter.

#### ShellFlyoutTemplatedContentRenderer

Modified:
* Updated the `LoadView` logic to properly insert the content from `FlyoutFooter`. 
* Once the content is inserted the `LoadView` logic calculates the `height` of the footer and applies the margin to the `RecyclerView`. This change ensures the `ShellContent` and `FlyoutItem` does not get rendered behind the `FlyoutFooter`

### Before/After Screenshots ### 
This is a new feature so there is no before screenshot. The after screenshot is a gif demonstrating the new feature.

shell_footer_demo

### Testing Procedure ###
I updated the `Xamarin.Forms.Sandbox` project to use the new feature and completed my testing there.

* Created Header and Footer
* Added FlyoutItems and ShellContent
* Added many items to force the `RecyclerView` to scroll and demonstrated the items are not hidden behind the `FlyoutFooter`

### PR Checklist ###

- [ ] Has automated tests 
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

#### Omitted Tests
There are currently no automated tests for this as the work focused on UI changes which can be difficult to write automation around. During the review process if we identify where tests are needed I am more than happy to add automated tests.

### Questions ###
While working on this PR I noticed that `ShellFlyoutContentRenderer` didn't seem to be used and the
 `ShellFlyoutTemplatedContentRender` render is used instead. Here is a code snippet that shows some stale code. I am not sure if this is a hold-over from an original attempt or needs to be removed or is there for a future implementation.

https://github.com/xamarin/Xamarin.Forms/blob/f0a33ee3d976b8ebf7dc9e899b2fd0d7c61e4d66/Xamarin.Forms.Platform.Android/Renderers/ShellRenderer.cs#L166-L170

### CODE REVIEW

**Comment by PureWeen** [comment]
@ahoefling were you planning to work on the iOS side of this as well at all?

**Comment by ahoefling** [comment]
I can, but I don't have a business need for iOS at the moment as my project is targeting Android and that is what I built this for. As of right now I am not planning to add an iOS portion for this PR unless it is required to move this PR forward.

I can certainly start looking at the iOS implementation sometime in the next week or so.

**Comment by PureWeen** [comment]
@ahoefling no worries on not working on iOS. I was only asking just so duplicate work didn't occur at all. This will get us to the Footer Templates a lot faster!! I have some work on ios I've been doing over here https://github.com/xamarin/Xamarin.Forms/tree/shell_fixes to smooth out the flyout so I'll probably just end up merging your work into that and then finishing up the footer on ios as well

We'll have to get an iOS implementation done before we can merge the Android one but this will get us there a lot faster!!

**Comment by ahoefling** [comment]
@PureWeen I'm glad this is going to make the implementation much easier and we will be able to use the work I started here. How should we proceed with getting this merged into your shell_fixes branch? Should I be expecting comments in a code review here or will you be taking ownership of this when you are ready to merge it into your branch?

Curious on our process forward with this.

**Comment by PureWeen** [comment]
@ahoefling 

> will you be taking ownership of this when you are ready to merge it into your branch?

Once the work for this is top of the queue then we'll just take what you've done here and add onto it. Since you've already done a lot of the guts then we should be able to move this up in priority a bit faster. 

I have a bunch of existing work I need to merge over on ios so I'm hoping I can take what you did here and easily provide a footer 🤞

**Comment by PureWeen** [comment]
@ahoefling Sorry about the delay but can you rebase this?
Let me know if you don't have time and I can take it over

**Comment by PureWeen** [comment]
Actually @ahoefling hold off on that request. 
I need to spike an alternate approach for this first

**Comment by ahoefling** [comment]
No problem, just let me know if I need to do anything

**Comment by samhouts** [comment]
I'm going to close this one for now until we're ready to return to it. Thanks, @ahoefling!

**Comment by ahoefling** [comment]
I spoke with @PureWeen about this earlier today and decided we would re-open the PR so we can get an implementation of `FlyoutFooter` into shell as opposed the best solution. I have re-implemented the changes based on 5.0.0 for Android as a starting point. I think there needs to be a little more work on how the Footer View is rendering as it is taking up too much space. This is a great starting point for us to finish up some work and get this in for 5.0.0

image

**Comment by ahoefling** [comment]
I fixed the footer height issue which was in the Xam Store implementation. I forgot to remove `HeightRequest` when I copied over the `FlyouHeader.xaml` to `FlyoutFooter.xaml`. Now the footer should be rendering correctly where it will wrap the height based on the `View`


image

**Review by PureWeen — APPROVED** [review]

**Review by pictos — APPROVED** [review]
LGTM!

**Pull request merged**
